test/unit/test_test_await_methods.gd

```diff
@@ -80,79 +87,86 @@ class TestTheNewWaitMethods:
-	func test_wait_until_is_compatible_with_checking_if_an_object_is_freed():
-		var node = add_child_autoqfree(Node.new())
-		# TODO: If we ever expose _utils.is_freed() in GUT's public API, we could swap it in here
-		# It makes the test pass without any error logs
-		# To add it to the API:
-		# 	- add it in addons/gut/test.gd
-		# 	- document in https://gut.readthedocs.io/en/latest/Asserts-and-Methods.html#utilities
-		# For now let's keep is_instance_valid to demo what a GUT user would probably use
-		var is_freed = func(): return not is_instance_valid(node)
+		signaler.emit_after(.5)
+		var result = await wait_for_signal(signaler.the_signal, 10)
+		assert_true(result)
+
+	func test_wait_for_signal_returns_false_when_signal_not_emitted():
 		var signaler = add_child_autoqfree(TimedSignaler.new())
-		signaler.the_signal.connect(node.queue_free)
-		signaler.emit_after(.1)
-
-		await assert_eventually(is_freed, 1.0)
-
-		assert_between(counter.time, 0.0, .2)
-		assert_freed(node)
-
-# ------------------------------------
-# Could not get these to trigger the error I was trying to replicate.  This was
-# a useful exercies though and I'm not ready to part with this code.  So it is
-# here as an example of using parameterized tests to control the order of the
-# execution of tests with different behavior per iteration.
-# ------------------------------------
-# class TestYieldTimerResetAfterSignalEmitted:
-# 	extends GutTest
-
-# 	var _wait_data = ParameterFactory.named_parameters(
-# 		['kind', 'time'],
-# 		[
-# 			['s', 3.0],
-# 			['t', 4.0],
-# 		])
-
-# 	var _counter = null
-# 	func before_each():
-# 		_counter = add_child_autoqfree(Counter.new())
-
-
-# 	func test_one(param=use_parameters(_wait_data)):
-# 		print('time left = ', gut._yield_timer.time_left)
-# 		var signaler = add_child_autoqfree(TimedSignaler.new())
-# 		var t = param.time
-
-# 		if(param.kind == 's'):
-# 			signaler.emit_after(t)
-# 			await wait_for_signal(signaler.the_signal, t + 2)
-# 		else:
-# 			await wait_seconds(t)
-
-# 		assert_between(_counter.time, t - 0.1, t + 0.1)
-# 		print('time left = ', gut._yield_timer.time_left)
\ No newline at end of file
+		signaler.emit_after(10)
+		var result = await wait_for_signal(signaler.the_signal, .5)
+		assert_false(result)
+
+
+	func test_wait_until_waits_ends_when_method_returns_true():
+		var all_is_good = func():
+			return counter.time > .25
+
+		await wait_until(all_is_good, .5)
+		assert_almost_eq(counter.time, .25, .05)
+		assert_false(did_wait_timeout(), 'did_wait_timeout')
+
+	func test_wait_until_times_out():
+		var all_is_good = func():
+			return false
+
+		await wait_until(all_is_good, .5)
+		assert_almost_eq(counter.time, .5, .05)
+		assert_true(did_wait_timeout(), 'did_wait_timeout')
+
+	func test_wait_until_returns_true_when_it_finishes():
+		var all_is_good = func():
+			return counter.time > .25
+
+		var result = await wait_until(all_is_good, .5)
+		assert_true(result)
```

Comment:
Pros:
- It opens the door to testing that "something won't happen"
  ```
  hero.strike_softly()
  assert_false(await wait_until(is_dead, 1))
  ````
- `wait_until` more coherent with the other `wait_*` functions

Cons:
- Easy to forget the assertion and mess up the test
  ```
  hero.strike_strongly() # Oops actually a soft hit
  await wait_until(is_dead, 1) # Oops it's not dead but still passes
  ```
- `wait_until(is_dead)` does not make it obvious if true or false is a "success". Maybe we could rephrase it as something like `eventually(is_dead)` which reads more easily in my mind
  - Btw English is my 2nd language, you can probably come up with a more natural phrasing haha

So I'm thinking the ideal world would be an `assert_eventually()` for most cases, and `eventually()` for rarer advanced use cases.
But it's really just syntax sugar at this point, nothing critical.

Reply:
> Easy to forget the assertion and mess up the test

GUT will mark any test that does not assert as risky so you know you didn't assert.  Also, when writing tests you should see them fail once before you implement the code that would make them pass.  Otherwise you can't know if you have tested anything (Red/Green/Refactor).

> wait_until(is_dead) does not make it obvious if true or false is a "success". 

The name of the test should give the `wait_until` the context it needs.  In this example I would expect the name to be something like `test_strong_strike_kills_player`.

We could add `wait_while` to do the inverse of `wait_until`.  This would wait until the predicate does not return `true`.  Then you could do something like `wait_while(is_alive)`.

Reply:
> The name of the test should give the wait_until the context it needs. In this example I would expect the name to be something like test_strong_strike_kills_player.

What I meant is that when you read"wait_until" it's not clear what the returned true or false mean. You have to read the docs (different libs use true as success and others false as success). So if the method name made it obvious it would be a plus. (Think is_alive => it's obvious that true means that the thing is alive, but if it was named check_alive, now you have to guess)

Reply:
> We could add wait_while to do the inverse of wait_until

Good idea!

If we had these 2, I can't think of a reason to wait until/while something and not assert on the result.
And if we always assert on the result, might as well re-add the assertion inside the wait_* methods.

Reply:
> GUT will mark any test that does not assert as risky so you know you didn't assert. Also, when writing tests you should see them fail once before you implement the code that would make them pass. Otherwise you can't know if you have tested anything (Red/Green/Refactor).

Totally agree and I always try to do that as much as possible. But that's an ideal world. The real world is messy, tests might have multiple assertions and GUT users will have varying levels of TDD skills

Reply:
> What I meant is that when you read"wait_until" it's not clear what the returned true or false mean

I understand now.  I had to take a step back and reassess, and it is a bit ambiguous.  My thought is that `wait_until(is_dead, x)` is returning what the predicate returned on the last call.   You would ignore "wait_until" when reading the line as English.  `if(await wait_until(is_dead, 10))` would be "If (wait a bit) is dead then".

Ok, ok, ok...you are onto something.  "wait a bit" is the same thing as "eventually".  I'm going to think on that a bit.  I may leave it as it is for now.  It would be easy to make syntax sugar or rename it and deprecate the old name.

> Good idea!

Cool.  I'm going to make an issue for `wait_while` so that this PR doesn't keep growing.

> If we had these 2, I can't think of a reason to wait until/while something and not assert on the result.

There still might be a reason to use these without asserts in integration tests.  For example, if you have a cooldown for attacks, you might use these to strike multiple times without having to worry that the script might hang forever.
```
player.hit(enemy)
wait_until(player.can_attack, 10)
player.hit(enemy)
wait_until(player.can_attack, 10)
player.hit(enemy)
assert_true(await wait_until(enemy.is_dead, 10))
```
Here, 10 is just some arbitrarily large value that you never expect to hit, not a time limit you are testing against.

> The real world is messy, tests might have multiple assertions and GUT users will have varying levels of TDD skills

I agree, it is really tough deciding scope.  In the end GUT is a TDD tool, so I have to treat it that way or I'll never stop tweaking things.  I really like tweaking things.

Reply:
When the behavior is not what you expected, wouldn't you want it to fail early?

It would give you more information
- The test did not fail because 3 hits failed to kill the enemy
- The test failed because the player was only able to hit 1 time when I wanted it to hit 3 times

```
player.hit(enemy)
wait_until(player.can_attack, 10) # Timeouts, dang it!
player.hit(enemy)
wait_until(player.can_attack, 10)
player.hit(enemy)
assert_true(await wait_until(enemy.is_dead, 10))
```

Reply:
This would be after you have tested the more simple interactions and you can assume that functions are working properly based on previous tests.  Probably an integration test.